File: mnheme_ports/java/mnheme/StorageEngine.java
// mnheme/java/mnheme/StorageEngine.java
// ======================================
// Append-only binary log engine.
// Frame: MAGIC(4B) | SIZE(4B big-endian) | JSON UTF-8 payload
// Deps: Jackson (com.fasterxml.jackson.core)

package mnheme;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.FileChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

public class StorageEngine {

    private static final byte[] MAGIC = {0x4D, 0x4E, 0x45, (byte) 0xE0};
    private static final int HEADER_SIZE = 8;

    private final Path path;
    private final ReentrantLock lock = new ReentrantLock();
    private final ObjectMapper mapper = new ObjectMapper();

    public StorageEngine(Path path) throws IOException {
        this.path = path;
        if (!Files.exists(path)) {
            Files.createFile(path);
        }
    }

    /**
     * Append a JSON record. Returns byte offset of the written record.
     */
    public long append(Map<String, Object> record) throws IOException {
        byte[] payload = mapper.writeValueAsBytes(record);
        int size = payload.length;

        ByteBuffer frame = ByteBuffer.allocate(HEADER_SIZE + size);
        frame.put(MAGIC);
        frame.order(ByteOrder.BIG_ENDIAN).putInt(size);
        frame.put(payload);
        frame.flip();

        lock.lock();
        try (FileChannel ch = FileChannel.open(
                path, StandardOpenOption.APPEND, StandardOpenOption.WRITE)) {
            long offset = ch.size();
            while (frame.hasRemaining()) ch.write(frame);
            ch.force(true); // fsync
            return offset;
        } finally {
            lock.unlock();
        }
    }

    /** Holds a scanned (offset, record) pair. */
    public record ScanResult(long offset, Map<String, Object> record) {}

    /**
     * Scan all valid records. Truncated/corrupted records are skipped silently.
     */
    @SuppressWarnings("unchecked")
    public List<ScanResult> scan() throws IOException {
        List<ScanResult> results = new ArrayList<>();
        if (!Files.exists(path)) return results;

        try (RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r")) {
            byte[] header = new byte[HEADER_SIZE];
            while (true) {
                long offset = raf.getFilePointer();
                int read = raf.read(header);
                if (read < HEADER_SIZE) break;

                if (!magicMatches(header)) {
                    raf.seek(offset + 1); // resync
                    continue;
                }

                int size = ByteBuffer.wrap(header, 4, 4)
                        .order(ByteOrder.BIG_ENDIAN).getInt();
                byte[] payload = new byte[size];
                if (raf.read(payload) < size) break; // truncated

                try {
                    Map<String, Object> rec = mapper.readValue(payload, Map.class);
                    results.add(new ScanResult(offset, rec));
                } catch (Exception ignored) { /* corrupted JSON */ }
            }
        }
        return results;
    }

    /**
     * Read a single record at a known offset — O(1).
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> readAt(long offset) throws IOException {
        try (RandomAccessFile raf = new RandomAccessFile(path.toFile(), "r")) {
            raf.seek(offset);
            byte[] header = new byte[HEADER_SIZE];
            if (raf.read(header) < HEADER_SIZE) return null;
            if (!magicMatches(header)) return null;

            int size = ByteBuffer.wrap(header, 4, 4)
                    .order(ByteOrder.BIG_ENDIAN).getInt();
            byte[] payload = new byte[size];
            if (raf.read(payload) < size) return null;

            try {
                return mapper.readValue(payload, Map.class);
            } catch (Exception e) {
                return null;
            }
        }
    }

    public long fileSize() throws IOException {
        return Files.size(path);
    }

    private static boolean magicMatches(byte[] header) {
        return header[0] == MAGIC[0] && header[1] == MAGIC[1]
                && header[2] == MAGIC[2] && header[3] == MAGIC[3];
    }
}

File: mnheme_ports/kotlin/StorageEngine.kt
// mnheme/kotlin/mnheme/StorageEngine.kt
// =======================================
// Append-only binary log engine — Android & JVM compatible.
// Zero external deps: uses only kotlin.stdlib + org.json (Android built-in).
//
// Frame: MAGIC(4B) | SIZE(4B big-endian) | JSON UTF-8 payload
//
// On Android: pass context.filesDir.resolve("mente.mnheme") as path.
// On JVM:     pass java.nio.file.Path or java.io.File.

package mnheme

import org.json.JSONObject
import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private val MAGIC       = byteArrayOf(0x4D, 0x4E, 0x45, 0xE0.toByte())
private const val HEADER_SIZE = 8

class StorageEngine(private val file: File) {

    private val lock = ReentrantLock()

    init {
        if (!file.exists()) file.createNewFile()
    }

    /**
     * Append a JSON record. Returns the byte offset of the written frame.
     */
    fun append(record: JSONObject): Long {
        val payload = record.toString().toByteArray(Charsets.UTF_8)
        val size    = payload.size

        val frame = ByteBuffer.allocate(HEADER_SIZE + size).apply {
            order(ByteOrder.BIG_ENDIAN)
            put(MAGIC)
            putInt(size)
            put(payload)
        }.array()

        return lock.withLock {
            file.outputStream().use { os ->
                val offset = file.length()
                os.channel.apply {
                    position(size()) // seek to end
                    write(ByteBuffer.wrap(frame))
                    force(true) // fsync
                }
                offset
            }
        }
    }

    data class ScanResult(val offset: Long, val record: JSONObject)

    /**
     * Scan all valid records. Truncated/corrupted records silently skipped.
     */
    fun scan(): List<ScanResult> {
        val results = mutableListOf<ScanResult>()
        if (!file.exists() || file.length() == 0L) return results

        RandomAccessFile(file, "r").use { raf ->
            val header = ByteArray(HEADER_SIZE)
            while (true) {
                val offset = raf.filePointer
                val read   = raf.read(header)
                if (read < HEADER_SIZE) break

                if (!header.take(4).toByteArray().contentEquals(MAGIC)) {
                    raf.seek(offset + 1) // resync
                    continue
                }

                val size = ByteBuffer.wrap(header, 4, 4)
                    .order(ByteOrder.BIG_ENDIAN).int
                val payload = ByteArray(size)
                if (raf.read(payload) < size) break // truncated

                runCatching { JSONObject(String(payload, Charsets.UTF_8)) }
                    .onSuccess { results.add(ScanResult(offset, it)) }
                // corrupted JSON: skip silently
            }
        }
        return results
    }

    /**
     * Read a single record at a known byte offset — O(1).
     */
    fun readAt(offset: Long): JSONObject? {
        if (!file.exists()) return null
        return runCatching {
            RandomAccessFile(file, "r").use { raf ->
                raf.seek(offset)
                val header = ByteArray(HEADER_SIZE)
                if (raf.read(header) < HEADER_SIZE) return null
                if (!header.take(4).toByteArray().contentEquals(MAGIC)) return null

                val size = ByteBuffer.wrap(header, 4, 4)
                    .order(ByteOrder.BIG_ENDIAN).int
                val payload = ByteArray(size)
                if (raf.read(payload) < size) return null
                JSONObject(String(payload, Charsets.UTF_8))
            }
        }.getOrNull()
    }

    fun fileSize(): Long = if (file.exists()) file.length() else 0L
}
